### PR TITLE
Fix Select focus when opening drop with search (#6901)

### PR DESCRIPTION
**Pull request by miqh**
Hello there,

These changes are intended to address #6901.

Happy to amend as necessary.

---

#### What does this PR do?

* Resolves the automatic double focus on the search input of Select described in the linked issue.
* Ensures the focus behaviour within the Select dropdown is consistent when the latter is opened via mouse click or by keyboard control. Currently, opening the dropdown via keyboard control will automatically focus the search input as well as the first option.

#### Where should the reviewer start?

Changes should be confined to files for the Select component.

#### What testing has been done on this PR?

Confirmed the changes manually through Storybook locally. I've also added two additional tests to cover the focus behaviours going forward.

#### How should this be manually tested?

Follow the steps outlined in the linked issue.

#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

(None of the existing tests for Select make use of `userEvent` and so I've followed the surroundings for consistency.)

#### Any background context you want to provide?

None.

#### What are the relevant issues?

#6901

#### Screenshots (if appropriate)

Keyboard control:

demo1

Mouse control:

demo2

#### Do the grommet docs need to be updated?

Unsure. Maintainers to decide.

#### Should this PR be mentioned in the release notes?

Unsure. Maintainers to decide.

#### Is this change backwards compatible or is it a breaking change?

Unlikely.

### CODE REVIEW

File: src/js/components/Select/__tests__/Select-test.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ import { Select } from '..';
 
 describe('Select', () => {
   window.scrollTo = jest.fn();
+  window.HTMLElement.prototype.scrollIntoView = jest.fn();
```
Comment (miqh) [inline src/js/components/Select/__tests__/Select-test.js:17]: I've relocated this here because the tests I've introduced are also affected by this function, which doesn't appear to exist in JSDOM.

I noticed the previous location made no attempt at restoring the original contents, though I suspect it's an inconsequential matter.

File: src/js/components/Select/SelectContainer.js [hunk 2]
```diff
@@ -136,7 +136,7 @@ const SelectContainer = forwardRef(
         const clearButton = clearRef.current;
         if (onSearch) {
           const searchInput = searchRef.current;
-          if (searchInput && searchInput.focus) {
+          if (searchInput && searchInput.focus && !activeRef.current) {
```
Comment (jcfilben) [inline src/js/components/Select/SelectContainer.js:139]: When I tested with only the change on line 118, it appeared to fix the issue so just wanted to know why this change was also necessary

**Review by jcfilben — COMMENTED** [review]
Thanks for adding a test!
Functionality looks good just a small question

**Comment by jcfilben** [comment]
You can fix the failing bundlesize test by increasing thebundleSize in the `package.json` file from 161 to 162.

```
  "bundlesize": [
    {
      "path": "./dist/grommet.min.js",
      "maxSize": "162 kB"
    }
  ],
```

File: src/js/components/Select/__tests__/Select-test.js [hunk 3]
```diff
@@ -1592,5 +1589,76 @@ describe('Select', () => {
...
+        <Select
+          id="test-select"
+          options={['one', 'two']}
+          onSearch={onSearch}
+          placeholder="test select"
+        />
+      </Grommet>,
+    );
+    const select = getByPlaceholderText('test select');
+    fireEvent.click(select);
```
Comment (taysea) [inline src/js/components/Select/__tests__/Select-test.js:1606]: @jcfilben Do you feel that this test is a candidate to use userEvent instead of fireEvent?

I know userEvent is preferred as a best practice but also has some shortcomings. Wanted to get your opinion before suggesting the PR author to make any changes.

Comment (jcfilben) [reply src/js/components/Select/__tests__/Select-test.js:1606]: Good question, I think places where we use useFakeTimers we've run into issues with `userEvent`. But there is a chance that `userEvent` could work here. I haven't nailed down the exact situations that it does and doesn't work I just know useFakeTimers + userEvent has caused issues on other tests

Comment (taysea) [reply src/js/components/Select/__tests__/Select-test.js:1606]: Thanks, Jessica. @miqh would you be open to trying `userEvent` in place of `fireEvent` in your tests? Let us know if you run into any issues related to the points Jessica mentioned above

Comment (miqh) [reply src/js/components/Select/__tests__/Select-test.js:1606]: @taysea, I've given it a shot, but to no ease of success.

The best I could arrive at was something like the following:

```js
// Test itself is marked async
const user = userEvent.setup();
...
const select = getByPlaceholderText('test select');
await user.click(select);
await waitFor(() => expect(getAllByRole('searchbox')[0]).toHaveFocus());
// Select does not respond to navigation if event is directed at the search input
await user.type(document.getElementById('test-select__select-drop'), '[ArrowDown]');
// Search input does unfocus
await waitFor(() => expect(getAllByRole('searchbox')[0]).not.toHaveFocus());
// But the focus is instead directed at a container element and not the first option
await waitFor(() => expect(getAllByRole('option')[0]).toHaveFocus());
```
I should also add that I tried `user.keyboard`, which fell short of even being able to shift focus (i.e. respond to the arrow down) once the drop appears.

This is presumably because the latter acts on the element in focus. However, as I've noted, dispatching the arrow down key while the search input is in focus causes the Select to not respond (i.e. something must be swallowing up that event?).

If I contrast this with my existing test, that single `fireEvent.keyDown` invocation, plus a delay, was able to correctly (or at least for the purpose of the test) simulate a keyboard navigation step and to also focus the first option as expected.

Any suggestions on your part? I'm not sure what to do at this stage if there's insistence on `userEvent` being used instead.

Otherwise I'll update the bundle size test and call it a day.

Comment (jcfilben) [reply src/js/components/Select/__tests__/Select-test.js:1606]: Hey @miqh thanks for trying `userEvent`, since it is not working I am okay with us sticking with the original test

File: src/js/components/Select/SelectContainer.js (hunk 2 above)
Comment (jcfilben) [reply src/js/components/Select/SelectContainer.js:139]: Ah gotcha, that make sense. Thanks for the explanation!

**Review by jcfilben — APPROVED** [review]
Looks good!

**Review by taysea — APPROVED** [review]
LGTM! Thank you for the contribution.
